### DISCUSSION TITLE
Add constants for property aliases to ModelsBuilder

### DISCUSSION THREAD

**Discussion by LennardF1989** [umbraco/Umbraco-CMS#13994] [Features and ideas]
Right now, the only way to sanely retrieve property aliases is by using the information in the `ImplementPropertyType`-attribute. This has a runtime overhead because you will have to use reflection to get to it.

Please add constants for all aliases like you do with `ModelTypeAlias`.

The use case of this would be when you are working with an `IContent`, instead of an `IPublishedContent`.

**Comment by abjerner** [comment]
Wouldn't `nameof(MyModel.MyProperty)` suffice?

Anyways - based on past comments from Umbraco HQ, it doesn't seem like they are going to add new features to the build-in Models Builder. But if you can come up a good example on how to expose the property aliases, I'd be happy to add something like this to my [custom Models Builder](https://github.com/limbo-works/Limbo.Umbraco.ModelsBuilder).

**Reply by leekelleher** [reply]
I've been using `nameof()` myself.  But that does give the PascalCase alias, not camelCase (as it's set in the CMS).  From what I can tell, Umbraco's content cache is case-insensitive, so no problem there.  Personally, I would like to have constants for the property aliases, but that's more about me being fussy.

**Reply by abjerner** [reply]
Yes, my experience is also that property aliases are case-insensitive. But I also tend to keep a set of constants for various property aliases. Right now I'm maintaining these manually, but I have thought about adding logic to auto-generate those as well.

Maybe a tool wouldn't be able to generate the constants exactly the way I like them (at least not without any configuration), but it could be a step on the way.

![image](https://user-images.githubusercontent.com/3634580/226645806-c2cedf20-5741-4783-85e4-431d12d3ed0c.png)

**Reply by LennardF1989** [reply]
But why go through great lengths to do all this manually, when for Umbraco it's two lines of code to do it for us:

https://github.com/umbraco/Umbraco-CMS/blob/33adbf41fa1f5c5d0759c70a7116114107addf56/src/Umbraco.Infrastructure/ModelsBuilder/Building/TextBuilder.cs#L347

```
        WriteGeneratedCodeAttribute(sb, "\t\t");
        sb.AppendFormat(
            "\t\tpublic new const string {0}Alias = \"{1}\";\n",
            property.ClrName,
            property.Alias);
```

I'll just PR this and see what happens.

**Reply by LennardF1989** [reply]
See #13995.

**Reply by leekelleher** [reply]
The only problem I see with #13995 is if someone added another property with an "Alias" suffix. e.g. `umbracoBytesAlias`, that'd cause a clash in the C# class. 🤔  _(I mean, who'd do such a thing? but once out in the wild, anything could happen)_

I'm wondering if they should be sub-classed, so to avoid clashes in the model itself? _(I'm spit balling here.)_

**Reply by LennardF1989** [reply]
Why? It will just stick Alias behind it :P `umbracoBytesAlias` will become `umbracoBytesAliasAlias`.

**Reply by LennardF1989** [reply]
Also see: https://github.com/umbraco/Umbraco-CMS/blob/ba801cd1944a1948a9ee887af7f6e88aab3f57d5/src/Umbraco.Infrastructure/ModelsBuilder/Building/TextBuilder.cs#L349

**Reply by leekelleher** [reply]
> Why? It will just stick Alias behind it :P `umbracoBytesAlias` will become `umbracoBytesAliasAlias`.

...and the alias for the original `umbracoBytes`? That'd be `const string UmbracoBytesAlias` and the new property would be the same name?

**Reply by LennardF1989** [reply]
My bad, typo. You are correct. I guess a prefix could work. I mean, technically you cant make a propertyAlias now with the name "ModelTypeAlias"... It doesn't have to be pretty, as long as it saves time and I get compile errors when trying to do weird stuff with aliases :P

Or the alternative, a flag to turn it on or off. Then if you turn this feature on, you are aware of the consequences.

**Reply by abjerner** [reply]
In my example with the manually maintained constants, I think it's easier to explore and access the various constants, as they can be grouped in sub classes they I think makes sense. These sub classes could align with content types, but it could also be something entirely different.

But I don't see my example and the constants in your PR as mutually exclusive. Maybe my example could use the constants from generated from something like in your PR? I both examples could be auto-generated.



> My bad, typo. You are correct. I guess a prefix could work. I mean, technically you cant make a propertyAlias now with the name "ModelTypeAlias"...

Maybe a sub class instead? Although that could technically also conflict 😱

**Reply by leekelleher** [reply]
@LennardF1989 ~The last line in the screengrab has a lowercase `u`.  ModelsBuilder would PascalCase this, right?~ _(ignore this, noticed you'd updated your reply)_

Sorry if I'm sounding like I'm nitpicking, I'm all for the feature, and thank you for taking the initiative on it. 👍

**Reply by leekelleher** [reply]
@LennardF1989 Good point about an property with the alias `modelTypeAlias`. Let's see what the Core Collaborators team say, if HQ don't see it as an issue, it should be fine. There are already rules in place for naming properties certain things (like "name"), so maybe the "Alias" suffix could be reserved? (I'm not strongly held on that opinion though)

**Reply by LennardF1989** [reply]
@abjerner Hence I think a feature flag is the only true way. It's backwards compatible, and if someone turns it on, they know the consequences in that you can get into situations where the alias can clash with other constants.

**Reply by LennardF1989** [reply]
Already added a opt-in for it @leekelleher :)

**Reply by bergmania** [reply]
Overall I like the idea :) 

I dont see any issues by adding this to ModelsBuilder directly in the CMS.

Fundamentally I'm not for more configurations as the test matrix increases in size every time we add a new configuration. If we add it to Umbraco 11 and potentially Umbraco 10, I will be required to avoid conflicts, but I could be tempted to enable it by default from Umbraco 12.

**Reply by LennardF1989** [reply]
I'm fine with either :) The option would be for backwards compatibility.

**Reply by skttl** [reply]
Why not have it generate the constants in another namespace, eg Umbraco.Cms.Web.Common.PublishedConstants.{ContentTypeAlias}.{PropertyAlias}

**Reply by LennardF1989** [reply]
While a good idea, the whole point of this commit is to make the least changes to the ModelsBuilder possible, under the idea of Umbraco HQ not really wanting to do much more with it. Generating a whole additional set of files would certainly be a much bigger change, not too mention a partial repeat of code due to the loop through all ContentTypes and properties.

**Reply by bielu** [reply]
@bergmania when I saw this thread, I made comments in PR, as I think it is not great in long run, because of potential Collisions when generating, it is why I propose alternative approach to wrap it with PropertyWrapper with properties Value and Alias and I think it is cleaner and easier to maintain solution in long run :)
as mentioned in pr from @LennardF1989 I made POC
https://github.com/umbraco/Umbraco-CMS/pull/14068
It didnt require much of changes, but it is POC I would need made same changes for properties coming from Mixins etc, but yeah it is just POC for now

**Comment by nul800sebastiaan** [comment]
Hey all! We've enjoyed and spent quite a bit of time all contemplating your feature wishes and the many varying opinions on what's the best solution. 

After listening to your wishes, edge cases and possible conflicts in generating classes we came to the following conclusion:

- the opinions on the concrete implementation here are quite varied
- we have too few unit tests for Modelsbuilder to confidently make changes in generating models
- a LOT of the CMS relies on Modelsbuilder staying stable

Therefore, we'd like these kinds of changes to be tested out in the wild in a good variety of websites to see what works and what needs fixing up before it's reliable.

To that end, we've merged #12210 that will allow anyone to implement their own `GenerateModels` method. We'd love it if one/some of you would make a package for generating models differently and have that used by various projects to stabilize the new additions. Once we're more confident that a certain implementation works reliably we can reconsider looking at a pull request to have these proven changes added to the core models generator.

We definitely realize that this means copying a lot of our code into that generator but in the end we think this is best for the stability of the CMS.

And finally a thanks to everyone for chiming in here, we love the engagement!
